File: src/Yass/Discover.hs
{-# LANGUAGE OverloadedStrings #-}

module Yass.Discover
  ( discoverSpecFiles
  , isYassYamlFile
  , relativizePath
  ) where

import Control.Exception (IOException, try)
import Data.List (sortBy)
import Data.Ord (comparing)
import qualified Data.Text as T
import qualified Data.Text.Normalize as Normalize
import qualified Data.ByteString as BS
import System.Directory
  ( doesFileExist
  , doesDirectoryExist
  , doesPathExist
  , listDirectory
  , pathIsSymbolicLink
  )
import System.FilePath ((</>), takeFileName, isAbsolute)
import qualified Data.Text.IO as TIO
import System.IO (hFlush, stderr)
import Yass.Types (YassError(..), ErrorCode(..))
import qualified Yass.ErrorLine as EL

-- | Check if a filename has the .yass.yaml suffix and a non-empty basename
-- before the suffix.  The name must not start with ".".
isYassYamlFile :: FilePath -> Bool
isYassYamlFile path =
  let name = takeFileName path
  in hasSuffix ".yass.yaml" name
     && length name > length (".yass.yaml" :: String)
     && head name /= '.'

-- | Check if a string ends with a given suffix (case-sensitive)
hasSuffix :: String -> String -> Bool
hasSuffix suffix str =
  let sLen = length suffix
      strLen = length str
  in strLen >= sLen && drop (strLen - sLen) str == suffix

-- | Relativize a path: make it relative to cwd, no leading "./", basename if
-- directly in cwd, forward slashes.  Delegates to ErrorLine.relativizePath.
relativizePath :: FilePath -> FilePath -> FilePath
relativizePath = EL.relativizePath

-- | Discover .yass.yaml spec files.
--
-- @discoverSpecFiles cwd projectRoot maybeInputPath@
--
-- If @maybeInputPath@ is @Just path@:
--   - If it's a file: verify .yass.yaml suffix and existence, return singleton
--   - If it's a directory: recursively find .yass.yaml files in it
--   - Otherwise: error
--
-- If @maybeInputPath@ is @Nothing@: search from @projectRoot@.
--
-- Results are sorted by Unicode code-point order on NFC-normalized paths.
discoverSpecFiles :: FilePath -> FilePath -> Maybe FilePath -> IO (Either [YassError] [FilePath])
discoverSpecFiles cwd projectRoot Nothing =
  discoverInDirectory cwd projectRoot
discoverSpecFiles cwd _projectRoot (Just inputPath) = do
  let absPath = if isAbsolute inputPath then inputPath
                else cwd </> inputPath
  -- Check if it's a symlink first
  isLink <- safeIsSymbolicLink absPath
  if isLink
    then do
      -- Symlink: check what the target is
      targetExists <- doesFileExist absPath  -- follows symlink
      if targetExists
        then handleFileArg cwd absPath
        else do
          isDir <- doesDirectoryExist absPath  -- follows symlink
          if isDir
            then handleDirArg cwd absPath
            else do
              pathExists <- doesPathExist absPath
              if pathExists
                then return $ Left [YassError
                  { yeFile    = Just (relativizePath cwd absPath)
                  , yeLine    = Nothing
                  , yeCode    = PathInvalidType
                  , yeMessage = "path is neither a file nor a directory: " <> T.pack absPath
                  }]
                else return $ Left [YassError
                  { yeFile    = Just (relativizePath cwd absPath)
                  , yeLine    = Nothing
                  , yeCode    = PathNotFound
                  , yeMessage = "path does not exist: " <> T.pack absPath
                  }]
    else do
      isFile <- doesFileExist absPath
      isDir  <- doesDirectoryExist absPath
      case (isFile, isDir) of
        (True, _) -> handleFileArg cwd absPath
        (_, True) -> handleDirArg cwd absPath
        _ -> do
          pathExists <- doesPathExist absPath
          if pathExists
            then return $ Left [YassError
                  { yeFile    = Just (relativizePath cwd absPath)
                  , yeLine    = Nothing
                  , yeCode    = PathInvalidType
                  , yeMessage = "path is neither a file nor a directory: " <> T.pack absPath
                  }]
            else return $ Left [YassError
                  { yeFile    = Just (relativizePath cwd absPath)
                  , yeLine    = Nothing
                  , yeCode    = PathNotFound
                  , yeMessage = "path does not exist: " <> T.pack absPath
                  }]

-- | Handle a file argument: verify it has .yass.yaml suffix and return singleton.
handleFileArg :: FilePath -> FilePath -> IO (Either [YassError] [FilePath])
handleFileArg cwd absPath =
  if isYassYamlFile absPath
    then do
      readable <- safeIsReadable absPath
      if readable
        then return $ Right [relativizePath cwd absPath]
        else return $ Left [YassError
              { yeFile    = Just (relativizePath cwd absPath)
              , yeLine    = Nothing
              , yeCode    = PathUnreadable
              , yeMessage = "cannot read path: " <> T.pack absPath
              }]
    else return $ Left [YassError
          { yeFile    = Just (relativizePath cwd absPath)
          , yeLine    = Nothing
          , yeCode    = PathBadExtension
          , yeMessage = "expected a .yass.yaml file: " <> T.pack absPath
          }]

-- | Handle a directory argument: check readability first, then discover.
-- Top-level directory args that are unreadable get PathUnreadable (not DiscoverDirUnreadable).
handleDirArg :: FilePath -> FilePath -> IO (Either [YassError] [FilePath])
handleDirArg cwd dirPath = do
  result <- try (listDirectory dirPath) :: IO (Either IOException [String])
  case result of
    Left _exc -> return $ Left [YassError
      { yeFile    = Just (relativizePath cwd dirPath)
      , yeLine    = Nothing
      , yeCode    = PathUnreadable
      , yeMessage = "cannot read path: " <> T.pack dirPath
      }]
    Right entries -> do
      (files, errs) <- walkDirectory cwd dirPath entries
      -- Emit errors for unreadable subdirectories to stderr
      mapM_ (emitDiscoverError cwd) errs
      let sorted = sortByNFC files
      return $ Right sorted

-- | Recursively discover .yass.yaml files in a directory.
discoverInDirectory :: FilePath -> FilePath -> IO (Either [YassError] [FilePath])
discoverInDirectory cwd dirPath = do
  result <- try (listDirectory dirPath) :: IO (Either IOException [String])
  case result of
    Left _exc -> return $ Left [YassError
      { yeFile    = Just (relativizePath cwd dirPath)
      , yeLine    = Nothing
      , yeCode    = PathUnreadable
      , yeMessage = "cannot read path: " <> T.pack dirPath
      }]
    Right entries -> do
      (files, errs) <- walkDirectory cwd dirPath entries
      -- Emit errors for unreadable subdirectories to stderr
      mapM_ (emitDiscoverError cwd) errs
      let sorted = sortByNFC files
      return $ Right sorted

-- | Walk a directory recursively, collecting .yass.yaml files.
-- Skips hidden directories and files, does not follow symlinks during traversal.
-- Returns (foundFiles, subdirectoryErrors).
walkDirectory :: FilePath -> FilePath -> [String] -> IO ([FilePath], [YassError])
walkDirectory cwd baseDir entries = do
  results <- mapM (processEntry cwd baseDir) entries
  let (fileLists, errLists) = unzip results
  return (concat fileLists, concat errLists)

-- | Process a single directory entry during recursive traversal.
-- Returns (foundFiles, errors) to propagate unreadable-subdirectory errors.
processEntry :: FilePath -> FilePath -> String -> IO ([FilePath], [YassError])
processEntry cwd baseDir entry
  -- Skip hidden entries (starting with ".")
  | head entry == '.' = return ([], [])
  | otherwise = do
      let fullPath = baseDir </> entry
      -- Check if it's a symlink - skip symlinks during traversal
      isLink <- safeIsSymbolicLink fullPath
      if isLink
        then return ([], [])
        else do
          isFile <- doesFileExist fullPath
          isDir  <- doesDirectoryExist fullPath
          case (isFile, isDir) of
            (True, _)
              | isYassYamlFile entry -> return ([relativizePath cwd fullPath], [])
              | otherwise            -> return ([], [])
            (_, True) -> do
              subResult <- try (listDirectory fullPath) :: IO (Either IOException [String])
              case subResult of
                Left _ -> return ([], [YassError
                  { yeFile    = Just (relativizePath cwd fullPath)
                  , yeLine    = Nothing
                  , yeCode    = DiscoverDirUnreadable
                  , yeMessage = "cannot read directory: " <> T.pack fullPath
                  }])
                Right sub -> walkDirectory cwd fullPath sub
            _ -> return ([], [])

-- | Emit a discover error to stderr, formatted as an ErrorLine.
emitDiscoverError :: FilePath -> YassError -> IO ()
emitDiscoverError cwd err = do
  let fp = maybe "yass" id (yeFile err)
  TIO.hPutStrLn stderr $ EL.formatErrorLineNoLine cwd fp (yeCode err) (yeMessage err)
  hFlush stderr

-- | Sort file paths by Unicode code-point order on NFC-normalized UTF-8 path.
sortByNFC :: [FilePath] -> [FilePath]
sortByNFC = sortBy (comparing nfcNormalize)
  where
    nfcNormalize :: FilePath -> T.Text
    nfcNormalize = Normalize.normalize Normalize.NFC . T.pack

-- | Safely check if a path is a symbolic link, returning False on error.
safeIsSymbolicLink :: FilePath -> IO Bool
safeIsSymbolicLink path = do
  result <- try (pathIsSymbolicLink path) :: IO (Either IOException Bool)
  case result of
    Left _  -> return False
    Right b -> return b

-- | Safely check if a file is readable.
safeIsReadable :: FilePath -> IO Bool
safeIsReadable path = do
  result <- try (BS.readFile path >> return True) :: IO (Either IOException Bool)
  case result of
    Left _  -> return False
    Right b -> return b

File: internal/shared/discover.go
package shared

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"sort"

	"golang.org/x/text/unicode/norm"

	yerrors "github.com/shakefu/yass/internal/errors"
)

// DiscoverSpecFiles finds .yass.yaml spec files based on the input path.
//
// When inputPath is a file: return that single file path (after checking extension).
// When inputPath is a directory: recursively find files with literal suffix ".yass.yaml".
// When inputPath is "": search recursively from projectRoot.
//
// Returns:
//   - filePaths: discovered file paths, formatted relative to cwd when possible
//   - nonFatalErrors: formatted error lines for yass.discover.dir_unreadable
//   - fatalError: fatal error (yass.path.not_found, yass.path.bad_extension, etc.)
func DiscoverSpecFiles(inputPath, cwd, projectRoot string) ([]string, []string, error) {
	// Normalize cwd and projectRoot to absolute paths with forward slashes.
	absCwd, err := filepath.Abs(cwd)
	if err != nil {
		return nil, nil, fmt.Errorf("resolving cwd: %w", err)
	}
	absCwd = filepath.Clean(absCwd)

	// Determine the effective path to use.
	effectivePath := inputPath
	if effectivePath == "" {
		effectivePath = projectRoot
	}

	// Make effective path absolute for stat.
	absPath := effectivePath
	if !filepath.IsAbs(absPath) {
		absPath = filepath.Join(absCwd, absPath)
	}
	absPath = filepath.Clean(absPath)

	// Lstat to get info without following symlinks.
	info, err := os.Lstat(absPath)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil, &PathError{
				Code:    yerrors.CodePathNotFound,
				Message: fmt.Sprintf(yerrors.Messages[yerrors.CodePathNotFound], effectivePath),
				Path:    effectivePath,
			}
		}
		if os.IsPermission(err) {
			return nil, nil, &PathError{
				Code:    yerrors.CodePathUnreadable,
				Message: fmt.Sprintf(yerrors.Messages[yerrors.CodePathUnreadable], effectivePath),
				Path:    effectivePath,
			}
		}
		return nil, nil, fmt.Errorf("stat %s: %w", effectivePath, err)
	}

	// If it's a symlink, stat to determine what it points to.
	resolvedInfo := info
	isSymlink := info.Mode()&fs.ModeSymlink != 0
	if isSymlink {
		resolvedInfo, err = os.Stat(absPath)
		if err != nil {
			if os.IsNotExist(err) {
				return nil, nil, &PathError{
					Code:    yerrors.CodePathNotFound,
					Message: fmt.Sprintf(yerrors.Messages[yerrors.CodePathNotFound], effectivePath),
					Path:    effectivePath,
				}
			}
			return nil, nil, &PathError{
				Code:    yerrors.CodePathUnreadable,
				Message: fmt.Sprintf(yerrors.Messages[yerrors.CodePathUnreadable], effectivePath),
				Path:    effectivePath,
			}
		}
	}

	// Classify: file or directory.
	if resolvedInfo.Mode().IsRegular() {
		// File input: check extension.
		if !hasYassSuffix(filepath.Base(absPath)) {
			return nil, nil, &PathError{
				Code:    yerrors.CodePathBadExtension,
				Message: fmt.Sprintf(yerrors.Messages[yerrors.CodePathBadExtension], effectivePath),
				Path:    effectivePath,
			}
		}
		formatted := formatDiscoverPath(absPath, absCwd)
		return []string{formatted}, nil, nil
	}

	if resolvedInfo.IsDir() {
		// Directory input: recursive search.
		files, nonFatal := discoverInDir(absPath, absCwd)
		return files, nonFatal, nil
	}

	// Neither file nor directory.
	return nil, nil, &PathError{
		Code:    yerrors.CodePathInvalidType,
		Message: fmt.Sprintf(yerrors.Messages[yerrors.CodePathInvalidType], effectivePath),
		Path:    effectivePath,
	}
}

// discoverInDir recursively searches a directory for .yass.yaml files.
// It does not follow symlinks, does not descend into hidden directories,
// and does not match hidden files.
func discoverInDir(dirPath, cwd string) ([]string, []string) {
	var files []string
	var nonFatalErrors []string

	walkDir(dirPath, cwd, &files, &nonFatalErrors)

	// Sort by Unicode code-point order on NFC-normalized path.
	sort.Slice(files, func(i, j int) bool {
		ni := norm.NFC.String(files[i])
		nj := norm.NFC.String(files[j])
		return ni < nj
	})

	return files, nonFatalErrors
}

// walkDir performs recursive directory walking without following symlinks.
func walkDir(dirPath, cwd string, files *[]string, nonFatalErrors *[]string) {
	entries, err := os.ReadDir(dirPath)
	if err != nil {
		// Non-fatal: emit dir_unreadable error line.
		formatted := formatDiscoverPath(dirPath, cwd)
		errLine := yerrors.FormatError(
			formatted,
			0,
			yerrors.CodeDiscoverDirUnreadable,
			fmt.Sprintf(yerrors.Messages[yerrors.CodeDiscoverDirUnreadable], formatted),
		)
		*nonFatalErrors = append(*nonFatalErrors, errLine)
		return
	}

	for _, entry := range entries {
		name := entry.Name()

		// Skip hidden entries.
		if isHidden(name) {
			continue
		}

		entryPath := filepath.Join(dirPath, name)

		// Check for symlinks during recursion: treat as absent.
		if entry.Type()&fs.ModeSymlink != 0 {
			continue
		}

		if entry.IsDir() {
			walkDir(entryPath, cwd, files, nonFatalErrors)
			continue
		}

		if entry.Type().IsRegular() {
			if matchesYassSpec(name) {
				formatted := formatDiscoverPath(entryPath, cwd)
				*files = append(*files, formatted)
			}
		}
	}
}

// matchesYassSpec checks if a filename is a valid .yass.yaml spec file:
// - ends with literal suffix ".yass.yaml"
// - has a non-empty basename before the suffix
// - does not start with "."
func matchesYassSpec(name string) bool {
	if isHidden(name) {
		return false
	}
	return hasYassSuffix(name)
}

// formatDiscoverPath formats a discovered path per the spec using
// errors.FormatPath:
// - relative when under cwd (no leading ./)
// - basename when directly in cwd
// - absolute otherwise
// Does not resolve symlinks.
func formatDiscoverPath(absPath, absCwd string) string {
	return yerrors.FormatPath(absPath, absCwd)
}
